### PR TITLE
Upload stories endpoint

### LINKED ISSUE #206: Story Upload - Backend
- add any fields that are currently missing (yt link, etc..)

- define the API for story upload  POST request
  - { 
       'video_url': yt_link,
       ....
     }

- adding logic to upload image that was given from frontend to S3 
   - look at @Shehryar21  pr to see how this is done
  
- create author in the author table if one does not already exist for the story

- To add tags to stories
   - how to store these in the backend (an array of tags per story)
  - keep a separate table in the backend for tags (one column is story id and second column is the tag -> primary key (story_id, tag))

- Convert regular youtube link to embed url (i.e https://www.youtube.com/watch?v=ao6jHx27gB8 -> https://www.youtube.com/embed/ao6jHx27gB8

Post Request Body Format
------------------------------------

Sending data as Form-Data
image: [file],
video_url: "",
title: "",
content: "",
summary: "",
author_first_name: "",
author_last_name: "",
author_country: "",
year: "",
current_city: "",
bio: "",
tags, []

Each tag would be added to another table with the story id and tag name.

### PR DESCRIPTION
**Pull request by Shehryar21**
Created a new endpoint (story). I made a new one instead of changing the existing stories endpoint. This adds a new story and also uploads image to s3. If author is not present, it creates the new author as well.
Partially closes #206

### CODE REVIEW

File: restapi/stories.go
Comment (AbhiPrasad) [inline restapi/stories.go]: lets move all constants outside of the endpoint. We can add them to the api interface in case other route handlers want access to them.

Constants: `awsAccessKeyID`, `awsSecretAccessKey`, `s3Config`

File: restapi/stories.go
Comment (AbhiPrasad) [inline restapi/stories.go]: Can we extract the s3 logic into a helper?

File: restapi/stories.go
Comment (AbhiPrasad) [inline restapi/stories.go]: Why the defer? It's a long function anyway so can we just close the file after it's uploaded?

File: restapi/stories.go
Comment (AbhiPrasad) [inline restapi/stories.go]: prefer `fmt.sprintf` over string concatenation with `+`

File: restapi/stories.go
Comment (AbhiPrasad) [inline restapi/stories.go]: Let's use a more descriptive variable then `i`?

File: restapi/stories.go
Comment (AbhiPrasad) [inline restapi/stories.go]: regex is constant, we can move it outside function.

File: restapi/stories.go
Comment (AbhiPrasad) [inline restapi/stories.go]: extract regex logic into helper function

File: restapi/stories.go
Comment (AbhiPrasad) [inline restapi/stories.go]: Move `https://www.youtube.com/embed/` into constant

**Review by AbhiPrasad — CHANGES_REQUESTED** [review]
1st pass review.

If you want anything clarified, DM me, we can hop on a call.

File: restapi/stories.go
Comment (nimwijetunga) [inline restapi/stories.go]: some of these fields can be optional (i.e they may not always appear in the api request) so you will need to make sure this logic still works

Comment (nimwijetunga) [reply restapi/stories.go]: Optional fields include: author bio, tags and youtube link

File: restapi/stories.go
Comment (nimwijetunga) [inline restapi/stories.go]: this logic should be wrapped in an if b/c the link may not always be present

File: restapi/stories.go
Comment (nimwijetunga) [inline restapi/stories.go:144]: this might be a common thing to do (i.e find an author given a struct) so maybe create a function in the models file that does this and call it here

File: restapi/stories.go
Comment (nimwijetunga) [inline restapi/stories.go]: are these keys already in heroku? If not add them please

**Comment by AbhiPrasad** [comment]
Also we should write table tests for this (https://github.com/golang/go/wiki/TableDrivenTests), but we can do that after this PR gets merged in.

File: restapi/api.go
Comment (nimwijetunga) [inline restapi/api.go]: Can you put this under PrivateAPI we don't want people to add stories until we finish auth

File: restapi/stories.go
Comment (nimwijetunga) [reply restapi/stories.go]: +1

**Review by nimwijetunga — CHANGES_REQUESTED** [review]

File: restapi/stories.go
Comment (Shehryar21) [reply restapi/stories.go]: image
you mean this api interface in api.go right?

File: restapi/stories.go
Comment (Shehryar21) [reply restapi/stories.go]: Is the Frontend doing any data validation making sure that link is a valid youtube link, or should i check in the backend

File: restapi/stories.go
Comment (AbhiPrasad) [reply restapi/stories.go]: yup

Comment (Shehryar21) [reply restapi/stories.go]: would it be better to just have s3Config as a constant since keyId and SecretKey are used in s3config anyways

File: restapi/stories.go
Comment (Shehryar21) [reply restapi/stories.go]: I remember we had a discussion where we might plan to change the filesize. In that case, wouldnt i need to defer it?

File: restapi/stories.go
Comment (AbhiPrasad) [reply restapi/stories.go]: yup, agree.

File: restapi/stories.go
Comment (AbhiPrasad) [reply restapi/stories.go]: Ah yeah true, I guess there is no harm in keeping it for now.

File: restapi/stories.go
Comment (AbhiPrasad) [reply restapi/stories.go]: we should always prioritize validation on the backend, never trust what a client sends you.

Being paranoid >>>>

File: restapi/stories.go
Comment (Shehryar21) [reply restapi/stories.go:144]: why in the models file?

File: restapi/stories.go
Comment (Shehryar21) [reply restapi/stories.go]: Some of them werent there (E.g bucket name), I added them

File: restapi/api.go
Comment (Shehryar21) [reply restapi/api.go]: Should I put under Private after Auth is done. Might be easier to test and integrate with Frontend cuz i believe we are removing old auth anyways

Comment (AbhiPrasad) [reply restapi/api.go]: wait why are we calling the endpoint `stories_formdata`? Shouldn't we just be POSTing on `stories`?

File: restapi/stories.go
Comment (AbhiPrasad) [inline restapi/stories.go:117]: I think we can make this a constant as well and move it outside of the function to the top of the file.

File: restapi/stories.go
Comment (AbhiPrasad) [inline restapi/stories.go]: Prefer `fmt.Sprintf`

File: restapi/stories.go
Comment (AbhiPrasad) [inline restapi/stories.go]: Let's name this `image_url` instead of res.

File: restapi/stories.go
Comment (AbhiPrasad) [inline restapi/stories.go]: nit: Let's call this function `uploadImageTos3`, make it more detailed

File: restapi/stories.go
Comment (AbhiPrasad) [inline restapi/stories.go]:
```suggestion
		return rest.ErrInvalidRequest(api.logger, "Error parsing year field", err)
```

File: restapi/stories.go
Comment (AbhiPrasad) [inline restapi/stories.go:125]: hmm, what happens if we get an array of stories in this endpoint? Shouldn't we be looping through the array?

**Review by AbhiPrasad — COMMENTED** [review]
Almost there :)

File: restapi/stories.go
Comment (nimwijetunga) [reply restapi/stories.go:144]: i feel like any methods where we query the model given some parameters should live in a specific file so they can be reused (I think the models file is the most sensible unless you can think of somewhere else?)

Comment (Shehryar21) [reply restapi/stories.go:144]: How about in the authors.go file since all the code related to authors is in that file?

File: restapi/api.go
Comment (Shehryar21) [reply restapi/api.go]: So i did not want to change the existing endpoint, which is why created a new one. Also the difference between these two is that `stories_formdata` (we can call it smth else or if we decide to get rid of the previous one, we can just call it stories) request payload is a formdata where as `stories` endpoint is a json. Also `stories_formdata` creates a single story, not sure how you would pass in a array of formdata. And looking at our use case, submitting a form on the Frontend would create just one story.

Comment (AbhiPrasad) [reply restapi/api.go]: I think we should just make the frontend send as `application/json` instead of form data, and make the endpoints JSON only.

Comment (Shehryar21) [reply restapi/api.go]: How would the FE pass in the image file in json (base64 encoding??) not sure how we would convert that in the backend but i can take a look at that if thats what we plan on doing. With form data, FE is just passing the image file directly.

Comment (AbhiPrasad) [reply restapi/api.go]: I mean the image file is just a byte array, so it can be accessed the same way in the JSON object iirc. 

I guess if it's too complicated we can leave it as is.

File: restapi/stories.go
Comment (Shehryar21) [reply restapi/stories.go:117]: I dont think regex.MustCompile can be initialized as a constant

File: restapi/api.go
Comment (Shehryar21) [reply restapi/api.go]: `file, h, err := r.FormFile("image")` This is how it reads the file from formdata. If its json, it would not be like this

File: restapi/stories.go
Comment (AbhiPrasad) [inline restapi/stories.go]:
```suggestion
	imageURL, err := api.uploadImageTos3(file, h.Size, h.Filename)
```

prefer lowercase here

**Review by AbhiPrasad — APPROVED** [review]
lgtm if it works after testing

**Review by nimwijetunga — APPROVED** [review]
lgtm (as long as it works after testing with Hanlin)

**Pull request merged**
